### PR TITLE
Add support for endpoint slices

### PR DESCRIPTION
**Pull request by clayton-gonsalves**
This PR adds support for Kubernetes endpoint slices. 

Fixes https://github.com/projectcontour/contour/issues/2696

### CODE REVIEW

File: internal/xdscache/v3/endpointslicetranslator.go
Comment (davinci26) [inline internal/xdscache/v3/endpointslicetranslator.go:41]: Do we want a prom stat counter here? I am thinking that it might be nice for debugging to be able to see the rate of recalculations so we can correlated with issues

**Review by davinci26 — COMMENTED** [review]
lgtm, one question and one test request

File: test/e2e/infra/endpointslice_test.go
Comment (clayton-gonsalves) [reply test/e2e/infra/endpointslice_test.go:62]: It would be hard to achieve a case where we can deterministically say during a test if the endpoints are ready or not. sure, w could remove the sleep function to catch some unready pods but that might lead to flakey tests

File: internal/xdscache/v3/endpointslicetranslator.go
Comment (clayton-gonsalves) [reply internal/xdscache/v3/endpointslicetranslator.go:46]: This is from a carry over from the endpoint slice implementation. I think we can get rid of this.

File: internal/xdscache/v3/endpointslicetranslator.go
Comment (clayton-gonsalves) [reply internal/xdscache/v3/endpointslicetranslator.go:41]: rechecking the code, there metrics should be there by default.

**Comment by clayton-gonsalves** [comment]
@skriss @sunjayBhatia  this is ready for the first pass

File: internal/xdscache/v3/endpointslicetranslator.go
Comment (skriss) [inline internal/xdscache/v3/endpointslicetranslator.go:227]: Let's just do a single DeepCopy() here, for the value.

File: internal/xdscache/v3/endpointslicetranslator.go
Comment (skriss) [inline internal/xdscache/v3/endpointslicetranslator.go:119]: The key is the NamespacedName of the _Service_, not the EndpointSlice -- let's note that here.

File: internal/xdscache/v3/endpointslicetranslator.go
Comment (skriss) [inline internal/xdscache/v3/endpointslicetranslator.go:40]: I'd say make this a method on EndpointSliceCache so we can name it just `RecalculateEndpoints` and it doesn't conflict with the existing one, since what it's recalculating is just (LoadBalancing) Endpoints.

File: internal/xdscache/v3/endpointslicetranslator.go
Comment (skriss) [inline internal/xdscache/v3/endpointslicetranslator.go:45]:
```suggestion
	for _, endpointSlice := range endpointSliceMap {
```

File: internal/xdscache/v3/endpointslicetranslator.go
Comment (skriss) [inline internal/xdscache/v3/endpointslicetranslator.go:214]:
```suggestion
// UpdateEndpointSlice adds endpointSlice to the cache, or replaces it if it is
```

File: internal/xdscache/v3/endpointslicetranslator.go
Comment (skriss) [inline internal/xdscache/v3/endpointslicetranslator.go:239]:
```suggestion
// DeleteEndpointSlice deletes endpointSlice from the cache. Any ServiceClusters
```

File: internal/xdscache/v3/endpointslicetranslator.go
Comment (skriss) [inline internal/xdscache/v3/endpointslicetranslator.go:242]:
```suggestion
func (c *EndpointSliceCache) DeleteEndpointSlice(endpointSlice *discoveryv1.EndpointSlice) bool {
```

**Review by skriss — CHANGES_REQUESTED** [review]
Thanks @clayton-gonsalves, this is looking pretty good - first pass just getting a few nits out of the way. Still looking into a few of the details

File: changelogs/unreleased/5745-clayton-gonsalves-minor.md
Comment (skriss) [reply changelogs/unreleased/5745-clayton-gonsalves-minor.md:1]:
```suggestion
## Add Kubernetes Endpoint Slice support
```

File: internal/xdscache/v3/endpointslicetranslator.go
Comment (skriss) [inline internal/xdscache/v3/endpointslicetranslator.go:64]: Still digesting the logic in this line, I feel like the `&&` should be an `||` (I know this matches what was in EndpointsTranslator though). We can address separately if it should be changed, since it would need to change in both if so.

**Review by skriss — CHANGES_REQUESTED** [review]

File: apis/projectcontour/v1alpha1/contourconfig.go
Comment (skriss) [inline apis/projectcontour/v1alpha1/contourconfig.go:91]: I'd be interested to hear what the other maintainers think about adding a more generic `FeatureFlags` field instead.  It could be used for other things, and deprecating a value for it should be simpler than deprecating an entire field.

Comment (clayton-gonsalves) [reply apis/projectcontour/v1alpha1/contourconfig.go:91]: I think thats a good idea, rather than polluting the config every time we want to add/deprecate a field.

File: internal/xdscache/v3/endpointslicetranslator.go
Comment (clayton-gonsalves) [reply internal/xdscache/v3/endpointslicetranslator.go:64]: yeah looking at it again, it feels like it should be an `||` because we only want to support TCP, we should short-circuit it if the protocol is anything else .

File: apis/projectcontour/v1alpha1/contourconfig.go
Comment (clayton-gonsalves) [reply apis/projectcontour/v1alpha1/contourconfig.go]: i made the change in https://github.com/projectcontour/contour/pull/5745/commits/1fd6eea12689e2de160d46537422f595af150d9b but will also wait for feedback from the other maintainers.

File: cmd/contour/serve.go
Comment (clayton-gonsalves) [reply cmd/contour/serve.go]: fixed in https://github.com/projectcontour/contour/pull/5745/commits/1fd6eea12689e2de160d46537422f595af150d9b

File: cmd/contour/serve.go
Comment (clayton-gonsalves) [reply cmd/contour/serve.go:604]: added in https://github.com/projectcontour/contour/pull/5745/commits/1fd6eea12689e2de160d46537422f595af150d9b

File: internal/xdscache/v3/endpointslicetranslator.go
Comment (clayton-gonsalves) [reply internal/xdscache/v3/endpointslicetranslator.go]: fixed in https://github.com/projectcontour/contour/pull/5745/commits/1fd6eea12689e2de160d46537422f595af150d9b

File: internal/xdscache/v3/endpointslicetranslator.go
Comment (clayton-gonsalves) [reply internal/xdscache/v3/endpointslicetranslator.go]: addressed in https://github.com/projectcontour/contour/pull/5745/commits/5f090766d8b3f26fbbb09740a202f797949c04df

File: internal/xdscache/v3/endpointslicetranslator.go
Comment (clayton-gonsalves) [reply internal/xdscache/v3/endpointslicetranslator.go:64]: I havent changed this yet, but happy to change this here and create an issue for `endpoints`

File: apis/projectcontour/v1alpha1/contourconfig.go
Comment (skriss) [reply apis/projectcontour/v1alpha1/contourconfig.go]: I was actually thinking a string slice, rather than a concrete field per feature, so that as features come and go it doesn't require any changes to the type itself, just to the values that are accepted in the field. But, would like to get a second opinion before making this change.

File: changelogs/unreleased/5745-clayton-gonsalves-minor.md
Comment (skriss) [inline changelogs/unreleased/5745-clayton-gonsalves-minor.md:4]: Let's add a sentence about this becoming the default down the road:

```suggestion
Note: This change is off by default and is gated by the feature flag `useEndpointSlice`.
This feature will be enabled by default in a future version on Contour once it has had sufficient bake time in production environments.
```

**Review by skriss — CHANGES_REQUESTED** [review]

File: internal/xdscache/v3/endpointslicetranslator.go
Comment (skriss) [reply internal/xdscache/v3/endpointslicetranslator.go]: I think it's desirable to sort, to produce stable results across runs. I forget the exact details but impacts performance and/or load balancing algorithms. So I'd say let's leave this in place.

**Comment by skriss** [comment]
Modulo the feature flags spec, this is looking good to me, no further comments at the moment

File: apis/projectcontour/v1alpha1/contourconfig.go
Comment (sunjayBhatia) [reply apis/projectcontour/v1alpha1/contourconfig.go]: +1 for string slice, similar in spirit to the `--disable-feature` flag which is more generic than adding a specific flag per thing we want to disable

I guess that naming makes it sound more generic than it is, it's really for disabling reconciling some optional extended support CRDs, i dunno if we should repurpose the existing flag, have to think about it a bit

Comment (skriss) [reply apis/projectcontour/v1alpha1/contourconfig.go]: Yeah, I was looking at that flag as well for possible use, but since its semantics are opt-out rather than opt-in, it seemed hard to use here (would you opt out of using Endpoints which would result in opting into EndpointSlices? Seems unintuitive). I also don't quite remember why we went with a flag vs. a config file/ContourConfig field for that, will have to look back. 

If we do just go ahead with a config file/ContourConfig field for opt-in features, then down the road we could always look at deprecating the `--disable-feature` flag and replacing with something more aligned with the opt-in feature flags.

File: changelogs/unreleased/5745-clayton-gonsalves-minor.md
Comment (clayton-gonsalves) [reply changelogs/unreleased/5745-clayton-gonsalves-minor.md:4]: addressed in https://github.com/projectcontour/contour/pull/5745/commits/ee4aa23bff2d152d469d36cf1d6ccf7dd8ce4777

File: cmd/contour/serve.go
Comment (clayton-gonsalves) [reply cmd/contour/serve.go]: addressed in https://github.com/projectcontour/contour/pull/5745/commits/ee4aa23bff2d152d469d36cf1d6ccf7dd8ce4777

File: apis/projectcontour/v1alpha1/contourconfig.go
Comment (clayton-gonsalves) [reply apis/projectcontour/v1alpha1/contourconfig.go]: added a string slice in https://github.com/projectcontour/contour/pull/5745/commits/ee4aa23bff2d152d469d36cf1d6ccf7dd8ce4777

is this what y'all had in mind?

File: internal/xdscache/v3/endpointslicetranslator.go
Comment (clayton-gonsalves) [reply internal/xdscache/v3/endpointslicetranslator.go]: just want to call out, that here the sorting happens only per `slice`, since we store all the slices in a map it's not possible to sort the combination of all the endpoints.

File: changelogs/unreleased/5745-clayton-gonsalves-minor.md
Comment (sunjayBhatia) [inline changelogs/unreleased/5745-clayton-gonsalves-minor.md:4]: nit
```suggestion
Note: This change is off by default and is gated by the feature flag `useEndpointSlices`.
```

File: apis/projectcontour/v1alpha1/contourconfig.go
Comment (sunjayBhatia) [inline apis/projectcontour/v1alpha1/contourconfig.go:89]: nit:
```suggestion
	// Available toggles are:
	// useEndpointSlices - Configures contour to fetch endpoint data
	// from k8s EndpointSlices instead of the default behavior which reads
	// endpoint data from Endpoint resources.
```

**Comment by skriss** [comment]
@clayton-gonsalves unfortunately looks like there are a couple small merge conflicts to resolve

File: apis/projectcontour/v1alpha1/contourconfig.go
Comment (skriss) [reply apis/projectcontour/v1alpha1/contourconfig.go]: Looks reasonable to me!

**Review by skriss — COMMENTED** [review]
No further comments from me, ready for this to go once the merge conflicts and latest comments are resolved

**Review by tsaarni — CHANGES_REQUESTED** [review]
Could you add the `featureFlags` also in Contour config file documentation here [`site/content/docs/main/configuration.md`](site/content/docs/main/configuration.md)? It often tends to get out-of-sync and the naming for attributes is not very consistent (dash vs camelcase). 

There is also example config in the same doc page, and same example config is in deployment manifest [`examples/contour/01-contour-config.yaml`](examples/contour/01-contour-config.yaml).

**Comment by clayton-gonsalves** [comment]
> Could you add the `featureFlags` also in Contour config file documentation here [`site/content/docs/main/configuration.md`](site/content/docs/main/configuration.md)? It often tends to get out-of-sync and the naming for attributes is not very consistent (dash vs camelcase).
> 
> There is also example config in the same doc page, and same example config is in deployment manifest [`examples/contour/01-contour-config.yaml`](examples/contour/01-contour-config.yaml).

I added it to the `configuration.md, I would prefer keeping the example configs as is because this feature is still opt in. I don't want folks copy pasting the config and enabling something by default.

**Comment by clayton-gonsalves** [comment]
@skriss @sunjayBhatia I think all comments are resolved. is this good to go?

**Review by tsaarni — COMMENTED** [review]
Just couple more questions from me, nothing important though!

File: internal/xdscache/v3/endpointslicetranslator.go
Comment (clayton-gonsalves) [reply internal/xdscache/v3/endpointslicetranslator.go:42]: It;s out of habit but I reason that a `map[T]struct{}` is intuitive as a set compared to a `map[T]bool`. A `map[T]bool` could have an element as `false` which doesn't make much sense and means you need to have an if check to be sure. 

If I am not mistaken, an empty struct also doesn't allocate memory. although at the scale we are using it, it won't make that big of a difference. If there is a precedence to use `map[T]bool` in this project I am happy to change it.

File: internal/xdscache/v3/endpointslicetranslator.go
Comment (clayton-gonsalves) [reply internal/xdscache/v3/endpointslicetranslator.go]: done

File: internal/xdscache/v3/endpointslicetranslator.go
Comment (clayton-gonsalves) [reply internal/xdscache/v3/endpointslicetranslator.go:48]: There is a discussion around it [here](https://github.com/projectcontour/contour/pull/5745#discussion_r1324600660),. 
I tried to follow the pattern from the endpoint implementation.

**Comment by skriss** [comment]
@tsaarni @sunjayBhatia anything else blocking here? I'd definitely like to get this into the upcoming 1.27 RC. Can always address minor stuff between RC and GA if needed.

**Review by skriss — APPROVED** [review]
LGTM, will get this into the upcoming 1.27 RC

**Comment by sunjayBhatia** [comment]
> @tsaarni @sunjayBhatia anything else blocking here? I'd definitely like to get this into the upcoming 1.27 RC. Can always address any remaining non-blocking items between RC and GA if needed.

nothing blocking from me 👍🏽

**Pull request merged**
